### PR TITLE
[glslang][extension] Add support for ARB_bindless_texture.

### PR DESCRIPTION
**Pull request by ShchchowAMD**
Add support for ARB_bindless_texture.

There's still some features related to non_uniform. Would update soon.

### CODE REVIEW

**Comment by greg-lunarg** [comment]
Can you please add tests for this functionality?

This can be done by adding shaders that use ARB_bindless_texture to the Test/ directory. You will also need to add the files to AST.FromFile.cpp. See switch.frag as an example.

You will also need to create result files in Test/baseResults/, but this can be done by building and running build/gtests/Release/glslangtests.exe --update-mode.

**Comment by ShchchowAMD** [comment]
> 

Thanks for reminding. I'll update test and out file here.

File: glslang/MachineIndependent/Versions.cpp [hunk 1]
```diff
@@ -227,6 +227,7 @@ void TParseVersions::initializeExtensionBehavior()
     extensionBehavior[E_GL_ARB_texture_query_lod]            = EBhDisable;
     extensionBehavior[E_GL_ARB_vertex_attrib_64bit]          = EBhDisable;
     extensionBehavior[E_GL_ARB_draw_instanced]               = EBhDisable;
+    extensionBehavior[E_GL_ARB_bindless_texture]          = EBhDisable;
```
Comment (greg-lunarg) [inline glslang/MachineIndependent/Versions.cpp:230]: Indentation

File: glslang/MachineIndependent/Versions.h [hunk 2]
```diff
@@ -163,6 +163,7 @@ const char* const E_GL_ARB_texture_query_lod            = "GL_ARB_texture_query_
 const char* const E_GL_ARB_vertex_attrib_64bit          = "GL_ARB_vertex_attrib_64bit";
 const char* const E_GL_ARB_draw_instanced               = "GL_ARB_draw_instanced";
 const char* const E_GL_ARB_fragment_coord_conventions   = "GL_ARB_fragment_coord_conventions";
+const char* const E_GL_ARB_bindless_texture          = "GL_ARB_bindless_texture";
```
Comment (greg-lunarg) [inline glslang/MachineIndependent/Versions.h:166]: Indentation

**Review by greg-lunarg — CHANGES_REQUESTED** [review]
Thanks for the test. Does your code support the full extension? If so, could we get tests for the other salient shader changes? Eg, from the extension spec:

`In particular, sampler and image types may be used as shader inputs/outputs, temporary variables, and uniform block members, and may be assigned to by shader code.`

Also, could we get an error message when this extension is used when targeting SPIR-V. For example, when compiled with this command line:

`glslangValidator.exe -V --target-env vulkan1.2 --amb --aml -o t.frag.spv t.frag`

The error would just say that this extension is not currently supported when targeting SPIR-V.

**Comment by ShchchowAMD** [comment]
Thanks, I'll make another PR today.

File: glslang/MachineIndependent/Versions.cpp (hunk 1 above)
Comment (ShchchowAMD) [reply glslang/MachineIndependent/Versions.cpp]: Fixed.

File: glslang/MachineIndependent/Versions.h (hunk 2 above)
Comment (ShchchowAMD) [reply glslang/MachineIndependent/Versions.h]: Fixed.

**Comment by ShchchowAMD** [comment]
> Thanks for the test. Does your code support the full extension? If so, could we get tests for the other salient shader changes? Eg, from the extension spec:
> 
> `In particular, sampler and image types may be used as shader inputs/outputs, temporary variables, and uniform block members, and may be assigned to by shader code.`
> 
> Also, could we get an error message when this extension is used when targeting SPIR-V. For example, when compiled with this command line:
> 
> `glslangValidator.exe -V --target-env vulkan1.2 --amb --aml -o t.frag.spv t.frag`
> 
> The error would just say that this extension is not currently supported when targeting SPIR-V.

Hi greg, thanks for reply. I've made another PR according to your reviews.

For tip2, I don't make sure whether I understand right, so just double check:

SPIR-V already has disassembled extension keyword:

`OpSourceExtension "GL_ARB_bindless_texture"`

So what we need to do is to disable usage of extension when generate SPV targeting vulkan?

Current glslang have checks to require spv version and vulkan, I add another check to allow some extensions reserved for vk in the latest PR.

Thanks for your review again.

**Comment by greg-lunarg** [comment]
Thank you for the error check, although I would like one additional small change :)

Can you please disallow for ALL SPIR-V, generating for OpenGL as well as Vulkan?

Specifically, could you change the name of the variable vkReservedExt to spvUnsupportedExt?

Also check instead spvVersion.spv and change Vulkan to SPIR-V in the error message?

**Comment by ShchchowAMD** [comment]
> 

Thanks for detailed suggestion. I've disabled it for spv and changed codes according to your comment.

**Review by greg-lunarg — APPROVED** [review]
LGTM. Thanks!

**Review by greg-lunarg — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**